core-api/src/main/java/com/optimizely/ab/optimizelyconfig/OptimizelyConfigService.java

```diff
@@ -15,89 +15,100 @@
  ***************************************************************************/
 package com.optimizely.ab.optimizelyconfig;
 
+import com.optimizely.ab.annotations.VisibleForTesting;
 import com.optimizely.ab.config.*;
 
 import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
 
 public class OptimizelyConfigService {
 
     private ProjectConfig projectConfig;
+    private OptimizelyConfig optimizelyConfig;
+    private static OptimizelyConfigService serviceInstance;
 
     // Map containing variables list for every feature key used for merging variation and feature variables.
     private Map<String, List<FeatureVariable>> featureKeyToVariablesMap;
 
-    /**
-     * create maps for experiment and features to be returned as one object
-     *
-     * @param projectConfig the current projectConfig
-     * @return {@link OptimizelyConfig} containing experiments and features
-     */
-    public OptimizelyConfig getOptimizelyConfig(ProjectConfig projectConfig) {
+    private OptimizelyConfigService(ProjectConfig projectConfig) {
         this.projectConfig = projectConfig;
         this.featureKeyToVariablesMap = generateFeatureKeyToVariablesMap();
 
         Map<String, OptimizelyExperiment> experimentsMap = getExperimentsMap();
-        return new OptimizelyConfig(
+        optimizelyConfig = new OptimizelyConfig(
             experimentsMap,
             getFeaturesMap(experimentsMap),
             projectConfig.getRevision()
         );
     }
 
+    /**
+     * creates new instance if the revision is updated
+     *
+     * @param projectConfig project config provided
+     * @return singleton object of {@link OptimizelyConfigService}
+     */
+    public static OptimizelyConfigService getInstance(ProjectConfig projectConfig) {
+        if (serviceInstance == null || !projectConfig.getRevision().equals(serviceInstance.getOptimizelyConfig().getRevision())) {
```

Comment:
We should check sdkKey value as well here since different projectConfig can have the same revision numbers accidentally.

Reply:
More important issue here is that we need to keep a cache per sdkKey, isn't it? 
We need a general cache management solution to support multiple OptConfig instances.